### PR TITLE
App Engine - SSL - Thrift Gateway - Cloud Bigtable

### PR DESCRIPTION
**Pull request by mbrukman**
This howto provides instructions for creating an App Engine app which will
connect to a load-balanced, autoscaled managed instance group of HBase Thrift
gateways which will connect to Cloud Bigtable.

Many thanks to Ben Menasha for the initial version of these instructions.

@bmenasha, @lesv — please review and comment.

### CODE REVIEW

**Comment by lesv** [comment]
A quick scan - it looks OK, I'll review in more detail in AM.

File: python/thrift/appengine-ssl-gateway/README.md [hunk 1]
```diff
@@ -0,0 +1,253 @@
+# Connecting from Google App Engine to Google Cloud Bigtable
+
+It's easy to deploy a Thrift gateway cluster on Google Compute Engine that
+proxies request to the Cloud Bigtable API.
+
+Here are step by step instructions for starting a Cloud Bigtable cluster and
+building and deploying a Thrift server on Compute Engine.
+
+**Table of contents**
```
Comment (lesv) [inline python/thrift/appengine-ssl-gateway/README.md]: While probably not a huge issue for the customers you are targeting this at, you should have a requirements section.  

`make` is on most systems, but not all, many Java folks don't use it.

File: python/thrift/appengine-ssl-gateway/thrift-gateway/deployment.jinja [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+          - key: google-container-manifest
+            value: |
+              version: v1
+              kind: Pod
+              metadata:
+                name: thrift-gateway
+              spec:
+                containers:
+                  - name: thrift-gateway
+                    image: gcr.io/{{ env['project'] }}/thrift-gateway
```
Comment (garye) [inline python/thrift/appengine-ssl-gateway/thrift-gateway/deployment.jinja]: Potential issues with project names with disallowed characters

File: python/thrift/appengine-ssl-gateway/.editorconfig [hunk 3]
```diff
@@ -0,0 +1,19 @@
+# For more info, see http://EditorConfig.org
```
Comment (lesv) [inline python/thrift/appengine-ssl-gateway/.editorconfig]: probably should be listed in a .gitignore file (and not be here)

Comment (mbrukman) [inline python/thrift/appengine-ssl-gateway/.editorconfig]: `.editorconfig` helps everyone contributing to a project have the same style, this is not a per-person configuration—it's not auto-generated and should be versioned, not ignored.

I think we should actually move this to the root directory of the repo but I didn't want to impose this style for all the files in the repo before checking with you and others.

I can take it out and we can submit this file separately as it's not crucial for this change.

File: python/thrift/appengine-ssl-gateway/README.md (hunk 1 above)
Comment (lesv) [inline python/thrift/appengine-ssl-gateway/README.md]: I just checked w/ our Python team, we don't deliver any samples w make.  If that's my only major comment however, I'll accept it, 'til we can do something more appropriate.

Comment (mbrukman) [inline python/thrift/appengine-ssl-gateway/README.md]: The alternative to using `make` is copy-pasting a bunch of longer command lines, with no error checking: e.g., if an environment variable is unset, you get a really inscrutable error which is hard to debug, as it doesn't actually say that the error is that the variable is being defaulted to the empty string.

If there's a better way of avoiding the rough edges for users, I'd love to use it, but this is the best I came up with to simplify the user experience and catch any errors (typos, forgetting to set env vars, etc.).

File: python/thrift/appengine-ssl-gateway/appengine/third_party/python/.gitignore [hunk 4]
```diff
@@ -0,0 +1 @@
+managed
```
Comment (lesv) [inline python/thrift/appengine-ssl-gateway/appengine/third_party/python/.gitignore]: This file can be much higher.

File: python/thrift/appengine-ssl-gateway/appengine/bigtable.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+class CreateTableHandler(webapp2.RequestHandler):
+    def get(self):
+        connection = CreateConnection()
+        table = 'mytable'
+        status = 'does not exist'
+        try:
+            connection.create_table(
+                'mytable',
+                {'cf1': dict(max_versions=10),
```
Comment (lesv) [inline python/thrift/appengine-ssl-gateway/appengine/bigtable.py]: 10 seems rather large.

File: python/thrift/appengine-ssl-gateway/thrift-gateway/install-hbase.sh [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    echo "INFO: skipping download of $(basename ${url}); already done."
+  fi
+}
+
+# Install the Cloud Bigtable connector.
+mkdir -p hbase-1.1.2/lib/bigtable
+
+download_url_to_file https://repo.maven.apache.org/maven2/io/netty/netty-tcnative/1.1.33.Fork9/netty-tcnative-1.1.33.Fork9-linux-x86_64.jar \
+    hbase-1.1.2/lib/bigtable/netty-tcnative-1.1.33.Fork9-linux-x86_64.jar
+download_url_to_file https://storage.googleapis.com/hadoop-lib/gcs/gcs-connector-latest-hadoop2.jar \
```
Comment (lesv) [inline python/thrift/appengine-ssl-gateway/thrift-gateway/install-hbase.sh]: Not needed for this application.

File: python/thrift/appengine-ssl-gateway/thrift-gateway/install-hbase.sh [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    hbase-1.1.2/lib/gcs-connector-latest-hadoop2.jar
+download_url_to_file http://repo1.maven.org/maven2/com/google/cloud/bigtable/bigtable-hbase-1.1/0.2.3/bigtable-hbase-1.1-0.2.3.jar \
```
Comment (lesv) [inline python/thrift/appengine-ssl-gateway/thrift-gateway/install-hbase.sh]: Really old version of the client -- many issues.

**Comment by lesv** [comment]
Just remove the .editorconfig, the rest are nits - your choice, LGTM

File: python/thrift/appengine-ssl-gateway/appengine/bigtable.py (hunk 5 above)
Comment (mbrukman) [inline python/thrift/appengine-ssl-gateway/appengine/bigtable.py]: Exactly, this is just a demo to create/list/delete tables, we don't actually store any data there.

File: python/thrift/appengine-ssl-gateway/thrift-gateway/install-hbase.sh (hunk 6 above)
Comment (mbrukman) [inline python/thrift/appengine-ssl-gateway/thrift-gateway/install-hbase.sh]: Good catch, removing; will upload new version shortly.

File: python/thrift/appengine-ssl-gateway/.editorconfig (hunk 3 above)
Comment (lesv) [inline python/thrift/appengine-ssl-gateway/.editorconfig]: We'll be enforcing style soon via other means, at least for Java.  Can you remove it from this PR, and do a separate one at the root.  I'll spend some time looking at it to see if we want to change things at all.

Comment (mbrukman) [inline python/thrift/appengine-ssl-gateway/.editorconfig]: Removing; will follow-up with a separate PR.

File: python/thrift/appengine-ssl-gateway/README.md (hunk 1 above)
Comment (mbrukman) [inline python/thrift/appengine-ssl-gateway/README.md]: Added a requirements section at the top.

File: python/thrift/appengine-ssl-gateway/appengine/third_party/python/.gitignore (hunk 4 above)
Comment (mbrukman) [inline python/thrift/appengine-ssl-gateway/appengine/third_party/python/.gitignore]: Combined with the `.gitignore` at the level of `appengine` directory.

File: python/thrift/appengine-ssl-gateway/thrift-gateway/deployment.jinja (hunk 2 above)
Comment (mbrukman) [inline python/thrift/appengine-ssl-gateway/thrift-gateway/deployment.jinja]: Fixed, thanks for catching this!

**Pull request merged**